src/compose/images.coffee

```diff
@@ -0,0 +1,301 @@
+Promise = require 'bluebird'
+_ = require 'lodash'
+EventEmitter = require 'events'
+logTypes = require '../lib/log-types'
+constants = require '../lib/constants'
+validation = require '../lib/validation'
+
+{ NotFoundError } = require '../lib/errors'
+
+# image = {
+# 	name: image registry/repo@digest or registry/repo:tag
+# 	appId
+# 	serviceId
+# 	serviceName
+# 	imageId (from resin API)
+# 	releaseId
+# 	dependent
+# 	dockerImageId
+# 	status Downloading, Downloaded, Deleting
+# 	downloadProgress
+# }
+
+hasDigest = (name) ->
+	name?.split?('@')?[1]? ? false
+
+module.exports = class Images extends EventEmitter
+	constructor: ({ @docker, @logger, @db }) ->
+		@imageCleanupFailures = {}
+		# A store of volatile state for images (e.g. download progress), indexed by imageId
+		@volatileState = {}
+
+	reportChange: (imageId, status) ->
+		if status?
+			@volatileState[imageId] ?= { imageId }
+			_.merge(@volatileState[imageId], status)
+			@emit('change')
+		else if imageId? and @volatileState[imageId]?
+			delete @volatileState[imageId]
+			@emit('change')
+
+	triggerFetch: (image, opts, onFinish) =>
+		onProgress = (progress) =>
+			@reportChange(image.imageId, { downloadProgress: progress.percentage })
+
+		onFinish ?= _.noop
+
+		@normalise(image.name)
+		.then (imageName) =>
+			image = _.clone(image)
+			image.name = imageName
+			@markAsSupervised(image)
+			.then =>
+				@inspectByName(imageName)
+				.then (img) =>
+					@db.models('image').update({ dockerImageId: img.Id }).where(image)
+			.then ->
+				onFinish(true)
+				return
+			.catch =>
+				@reportChange(image.imageId, _.merge(_.clone(image), { status: 'Downloading', downloadProgress: 0 }))
+				Promise.try =>
+					if validation.checkTruthy(opts.delta) and opts.deltaSource? and opts.deltaSource != 'resin/scratch'
+						@logger.logSystemEvent(logTypes.downloadImageDelta, { image })
+						@inspectByName(opts.deltaSource)
+						.then (srcImage) =>
+							opts.deltaSourceId = srcImage.Id
+							@docker.rsyncImageWithProgress(imageName, opts, onProgress)
+							.tap (id) =>
+								if !hasDigest(imageName)
+									@docker.getRepoAndTag(imageName)
+									.then ({ repo, tag }) =>
+										@docker.getImage(id).tag({ repo, tag })
+					else
+						@logger.logSystemEvent(logTypes.downloadImage, { image })
+						@docker.fetchImageWithProgress(imageName, opts, onProgress)
+				.then (id) =>
+					@db.models('image').update({ dockerImageId: id }).where(image)
+				.then =>
+					@logger.logSystemEvent(logTypes.downloadImageSuccess, { image })
+					return true
+				.catch (err) =>
+					@logger.logSystemEvent(logTypes.downloadImageError, { image, error: err })
+					return false
+				.then (success) =>
+					@reportChange(image.imageId)
+					onFinish(success)
+					return
```

Comment:
useless return

Reply:
Same as above

Reply:
In that case it should be `return null`: http://bluebirdjs.com/docs/warning-explanations.html#warning-a-promise-was-created-in-a-handler-but-was-not-returned-from-it